src/v/datalake/catalog_schema_manager.cc

```diff
@@ -168,49 +167,81 @@ simple_schema_manager::get_table_info(
 ss::future<checked<std::nullopt_t, schema_manager::errc>>
 catalog_schema_manager::ensure_table_schema(
   const iceberg::table_identifier& table_id,
-  const iceberg::struct_type& desired_type,
+  const iceberg::struct_type& writer_struct_type,
   const iceberg::unresolved_partition_spec& partition_spec) {
     auto gh = maybe_gate();
     if (gh.has_error()) {
         co_return gh.error();
     }
     auto load_res = co_await catalog_.load_or_create_table(
-      table_id, desired_type, partition_spec);
+      table_id, writer_struct_type, partition_spec);
     if (load_res.has_error()) {
         co_return log_and_convert_catalog_err(
           load_res.error(), fmt::format("Error loading table {}", table_id));
     }
 
     iceberg::transaction txn(std::move(load_res.value()));
+    const auto* schema = txn.table().get_schema(txn.table().current_schema_id);
 
-    // Check schema compatibility
-    auto type_copy = desired_type.copy();
-    auto get_res = get_ids_from_table_meta(table_id, txn.table(), type_copy);
-    if (get_res.has_error()) {
-        co_return get_res.error();
-    }
+    auto writer_struct_type_with_ids = writer_struct_type.copy();
+    auto fill_res = iceberg::try_fill_field_ids(
+      schema->schema_struct, writer_struct_type_with_ids);
+    if (fill_res == iceberg::ids_filled::no) {
+        auto merged_schema_struct_type = schema->schema_struct.copy();
+        auto merge_res = iceberg::merge_struct_types(
+          writer_struct_type, merged_schema_struct_type);
+        if (merge_res.has_error()) {
+            vlog(
+              datalake_log.warn,
+              "Failed to merge record schema with table schema {}: {}",
+              table_id,
+              merge_res.error());
+            co_return errc::not_supported;
+        }
+
+        // Make a copy so that we still have the original struct in case we need
+        // to log it on the error path.
```

Comment:
If we're only copying on the error case for logging, can we copy _only_ if we error? E.g. reapply `iceberg::merge_struct_types(writer_struct_type, merged_schema_struct_type)` in the error case? It feels pretty wasteful otherwise

Reply:
99% of the cases we won't go down this path/won't need schema evolution. I prefer this code. Easier on the eye.